File: 1325-path-with-maximum-probability/1325-path-with-maximum-probability.go
type Edge struct {
	node int
	prob float64
}

type MaxHeap []Edge

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i].prob > h[j].prob }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MaxHeap) Push(x interface{}) {
	*h = append(*h, x.(Edge))
}

func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func maxProbability(n int, edges [][]int, succProb []float64, start_node int, end_node int) float64 {
	graph := make([][]Edge, n)
	for i := 0; i < len(edges); i++ {
		a := edges[i][0]
		b := edges[i][1]
		prob := succProb[i]
		graph[a] = append(graph[a], Edge{node: b, prob: prob})
		graph[b] = append(graph[b], Edge{node: a, prob: prob})
	}

	prob := make([]float64, n)
	prob[start_node] = 1.0

	maxHeap := &MaxHeap{}
	heap.Init(maxHeap)
	heap.Push(maxHeap, Edge{node: start_node, prob: 1.0})

	for maxHeap.Len() > 0 {
		curr := heap.Pop(maxHeap).(Edge)
		currNode := curr.node
		currProb := curr.prob

		if currNode == end_node {
			return currProb
		}

		for _, next := range graph[currNode] {
			nextNode := next.node
			nextProb := currProb * next.prob

			if nextProb > prob[nextNode] {
				prob[nextNode] = nextProb
				heap.Push(maxHeap, Edge{node: nextNode, prob: nextProb})
			}
		}
	}

	return 0.0
}

File: 1325-path-with-maximum-probability/1325-path-with-maximum-probability.cpp
class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start_node, int end_node) {
        vector<vector<pair<int, double>>> graph(n);
        for(int i = 0; i < edges.size(); i++) {
            int a = edges[i][0];
            int b = edges[i][1];
            graph[a].push_back({b, succProb[i]});
            graph[b].push_back({a, succProb[i]});
        }

        queue<pair<int, double>> q;
        q.push({start_node, 1.0});
        vector<double> prob(n, 0.0);
        prob[start_node] = 1.0;

        while(!q.empty()) {
            int currNode = q.front().first;
            double currProb = q.front().second;
            q.pop();

            for(const auto &next : graph[currNode]) {
                int nextNode = next.first;
                double nextProb = next.second * currProb;

                if(nextProb > prob[nextNode]) {
                    prob[nextNode] = nextProb;
                    q.push({nextNode, nextProb});
                }
            }
        }

        return prob[end_node];
    }
};